keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,552 @@
+|Timer trees| Instead of ordered list use a sorted binary tree. | More efficient. Insertion is O(log n) |
+|Heap based timer|A variant of ordered timer list where heap is used to store the next expiry time | Efficient compared to ordered list. Bookkeeping and insertion is O(log n). |
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
+
+
+We will use a use a separate [`DelayingInterface`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L37) which has a heap based single shot api [`AddAfter`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L150). Every time we process an entry from this queue, we would need to add it back to the queue to simulate a periodic timer.
+
+
+For further reading:
+1. [Reinventing timer wheel](https://lwn.net/Articles/646950/)
+2. [Hashed and hierarchical timer wheel](http://www.cs.columbia.edu/~nahum/w6998/papers/sosp87-timing-wheels.pdf)
+3. [Golang timers in multi-cpu systems](https://github.com/golang/go/commit/76f4fd8a5251b4f63ea14a3c1e2fe2e78eb74f81)
+4. [Go timerwheel](https://github.com/RussellLuo/timingwheel)
+
+##### Metrics
+We propose to add metrics that could expose the performance health of the controller including and not limited to: skew, queue depth, job failures, job successes etc.
+
+
+### Add .status.lastSuccessfulTime
+[#issue/75674](https://github.com/kubernetes/kubernetes/issues/75674)  
+Add `lastSuccessfulTime` to `.status` that tracks the last time the job completed successfully. This will augment the `lastScheduledTime` available in the `.status` in the v1beta1 api. Potential use is in monitoring (e.g. fire an alert if lastSuccessfulTime is more than X ago).
+
+Code for reference, subject to change:
+
+```golang
+func (cjc *CronJobController)  syncOne(sj *batchv1beta1.CronJob, js []batchv1.Job, now time.Time, jc jobControlInterface, sjc sjControlInterface, recorder record.EventRecorder) {
+   ...
+   // Get most recent successfully finished job from current job list.
+   sj.Status.LastSuccessfulTime = &metav1.Time{Time: scheduledTime}
+	if _, err := sjc.UpdateStatus(sj); err != nil {
+		klog.Infof("Unable to update status for %s (rv = %s): %v", nameForLog, sj.ResourceVersion, err)
+	}
+   ...
+}
+```
```

Comment:
No need for code, KEPs are for expressing intent.

Reply:
ack.